src/game/engine/basec.cpp

```diff
@@ -0,0 +1,205 @@
+/**
+ * @file
+ *
+ * @author tomsons26
+ *
+ * @brief
+ *
+ * @copyright Chronoshift is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "basec.h"
+#include "gameini.h"
+#include "house.h"
+#include "iomap.h"
+#include "pipe.h"
+#include "straw.h"
+#include <stdio.h>
+
+#ifndef GAME_DLL
+BaseClass g_Base;
+#endif
+
+/**
+ * @brief
+ *
+ * @address 0x004265F8
+ */
+const BuildingClass *BaseClass::Get_Building(int entry) const
+{
+    BuildingClass *ptrs[7];
+    const BaseNodeClass &node = m_Nodes[entry];
+    CellClass &cell = Map[node.m_Cell];
+
+    ptrs[0] = cell.Cell_Building();
+    int count = 1;
+    for (int i = 0; i < OVERLAPPER_COUNT; ++i) {
+        if (cell.Get_Overlapper(i) != nullptr) {
+            ptrs[count++] = (BuildingClass *)cell.Get_Overlapper(i);
+        }
+    }
+
+    for (int j = 0; j < count; ++j) {
+        if (ptrs[j] != nullptr && ptrs[j]->Get_Cell() == node.m_Cell) {
+            if (ptrs[j]->What_Am_I() == RTTI_BUILDING && ptrs[j]->What_Type() == node.m_Type) {
+                return ptrs[j];
+            }
+        }
+    }
+    return nullptr;
+}
+
+/**
+ * @brief
+ *
+ * @address 0x0042674C
+ */
+BaseNodeClass *BaseClass::Get_Node(BuildingClass *building)
+{
+    for (int i = 0; i < m_Nodes.Count(); ++i) {
+        if (m_Nodes[i].m_Type == building->What_Type() && building->Get_Cell() == m_Nodes[i].m_Cell) {
+            return &m_Nodes[i];
+        }
+    }
+    return nullptr;
+}
+
+/**
+ * @brief
+ *
+ * @address 0x004267C4
+ */
+BaseNodeClass *BaseClass::Get_Node(cell_t cellnum)
+{
+    for (int i = 0; i < m_Nodes.Count(); ++i) {
+        if (m_Nodes[i].m_Cell == cellnum) {
+            return &m_Nodes[i];
+        }
+    }
+    return nullptr;
+}
+
+/**
+ * @brief
+ *
+ * @address 0x004267F8
+ */
+BaseNodeClass *BaseClass::Next_Buildable(BuildingType type)
+{
+    for (int i = 0; i < m_Nodes.Count(); ++i) {
+        if (type == BUILDING_NONE) {
+            if (!Is_Built(i)) {
+                return &m_Nodes[i];
+            }
+        } else {
+            if (type == m_Nodes[i].m_Type && !Is_Built(i)) {
+                return &m_Nodes[i];
+            }
+        }
+    }
+    return nullptr;
+}
+
+/**
+ * @brief
+ *
+ * @address 0x00426470
+ */
+BOOL BaseClass::Load(Straw &straw)
+{
+    int size = 0;
+    int count = 0;
+    BaseNodeClass node;
+
+    if (straw.Get(&size, sizeof(size)) != sizeof(size)) {
```

Comment:
Use fixed width types like int32_t for anything that gets saved and loaded, that includes member variables. We need to apply this more rigorously in general to ensure we don't have problems down the road.